stdlib/mqtt/to.go

```diff
@@ -0,0 +1,649 @@
+	Type flux.ColType
+}
+
+func (t *ToMQTTTransformation) Process(id execute.DatasetID, tbl flux.Table) error {
+	// set up the MQTT options.
+	opts := MQTT.NewClientOptions().AddBroker(t.spec.Spec.Broker)
+	if len(t.spec.Spec.ClientID) > 0 {
+		opts.SetClientID(t.spec.Spec.ClientID)
+	} else {
+		opts.SetClientID("flux-mqtt")
+	}
+	if t.spec.Spec.Timeout > 0 {
+		opts.SetConnectTimeout(t.spec.Spec.Timeout)
+	}
+	if len(t.spec.Spec.Username) > 0 {
+		opts.SetUsername(t.spec.Spec.Username)
+	}
+	if len(t.spec.Spec.Password) > 0 {
+		opts.SetPassword(t.spec.Spec.Password)
+	}
+	mqttTopic := t.spec.Spec.Topic
+
+	client := MQTT.NewClient(opts)
+	if len(t.spec.Spec.Message) > 0 {
+		
+		//create and start a client using the above ClientOptions
+		if token := client.Connect(); token.Wait() && token.Error() != nil {
+			return token.Error()
+		}
+  		token := client.Publish(t.spec.Spec.Topic, 0, false, t.spec.Spec.Message)
+  		token.Wait()
+		client.Disconnect(250)
+		return nil
+	}
+	pr, pw := io.Pipe() // TODO: replce the pipe with something faster
+	m := &toMqttMetric{}
+	e := protocol.NewEncoder(pw)
+	e.FailOnFieldErr(true)
+	e.SetFieldSortOrder(protocol.SortFields)
+	cols := tbl.Cols()
+	labels := make(map[string]idxType, len(cols))
+	for i, col := range cols {
+		labels[col.Label] = idxType{Idx: i, Type: col.Type}
+	}
+
+	// do time
+	timeColLabel := t.spec.Spec.TimeColumn
+	timeColIdx, ok := labels[timeColLabel]
+
+	if !ok {
+		return errors.New("Could not get time column")
+	}
+	if timeColIdx.Type != flux.TTime {
+		return fmt.Errorf("column %s is not of type %s", timeColLabel, timeColIdx.Type)
+	}
+	var measurementNameCol string
+	if t.spec.Spec.Name == "" {
+		measurementNameCol = t.spec.Spec.NameColumn
+	}
+
+	// check if each col is a tag or value and cache this value for the loop
+	colMetadatas := tbl.Cols()
+	isTag := make([]bool, len(colMetadatas))
+	isValue := make([]bool, len(colMetadatas))
+
+	for i, col := range colMetadatas {
+		valIdx := sort.SearchStrings(t.spec.Spec.ValueColumns, col.Label)
+		isValue[i] = valIdx < len(t.spec.Spec.ValueColumns) && t.spec.Spec.ValueColumns[valIdx] == col.Label
+
+		tagIdx := sort.SearchStrings(t.spec.Spec.TagColumns, col.Label)
+		isTag[i] = tagIdx < len(t.spec.Spec.TagColumns) && t.spec.Spec.TagColumns[tagIdx] == col.Label
+	}
+
+	builder, new := t.cache.TableBuilder(tbl.Key())
+	if new {
+		if err := execute.AddTableCols(tbl, builder); err != nil {
+			return err
+		}
+	}
+
+	var wg syncutil.WaitGroup
+	wg.Do(func() error {
+		m.name = t.spec.Spec.Name
+		err := tbl.Do(func(er flux.ColReader) error {
+			l := er.Len()
+			for i := 0; i < l; i++ {
+				m.truncateTagsAndFields()
+				for j, col := range er.Cols() {
+					switch {
+					case col.Label == timeColLabel:
+						m.t = values.Time(er.Times(j).Value(i)).Time()
+					case measurementNameCol != "" && measurementNameCol == col.Label:
+						if col.Type != flux.TString {
+							return errors.New("invalid type for measurement column")
```

Comment:
these types of errors should be checked outside of tbl.Do  using `tbl.Cols()`.  Doing this check over and over for every row will slow you down quite a bit.

Reply:
Sounds like we should leave this as-is.